src/db/mysql/batch.rs

```diff
@@ -0,0 +1,272 @@
+    // skew the sharding distribution.
+    //
+    // So we mix in the lowest digit of the uid to improve the distribution
+    // while still letting us treat these ids as millisecond timestamps.  It's
+    // yuck, but it works and it keeps the weirdness contained to this single
+    // line of code.
+    let id = timestamp + (user_id % 10) as i64;
+    let bsos = bsos_to_batch_string(&params.bsos)?;
+    insert_into(batches::table)
+        .values((
+            batches::user_id.eq(&user_id),
+            batches::collection_id.eq(&collection_id),
+            batches::id.eq(&id),
+            batches::bsos.eq(&bsos),
+            batches::expiry.eq(timestamp + BATCH_LIFETIME),
+        )).execute(&db.conn)?;
+    Ok(id)
+}
+
+pub fn validate(db: &MysqlDb, params: params::ValidateBatch) -> Result<bool> {
+    let user_id = params.user_id.legacy_id as i32;
+    let collection_id = db.get_collection_id(&params.collection)?;
+    let exists = batches::table
+        .select(sql::<Integer>("1"))
+        .filter(batches::user_id.eq(&user_id))
+        .filter(batches::collection_id.eq(&collection_id))
+        .filter(batches::id.eq(&params.id))
+        .get_result::<i32>(&db.conn)
+        .optional()?;
+    Ok(exists.is_some())
+}
+
+pub fn append(db: &MysqlDb, params: params::AppendToBatch) -> Result<()> {
+    // XXX: this is possibly an upsert
+    let user_id = params.user_id.legacy_id as i32;
+    let collection_id = db.get_collection_id(&params.collection)?;
+    let bsos = bsos_to_batch_string(&params.bsos)?;
+    // XXX: also update expiry?
+    update(batches::table)
+        .filter(batches::user_id.eq(&user_id))
+        .filter(batches::collection_id.eq(&collection_id))
+        .filter(batches::id.eq(&params.id))
+        .set(batches::bsos.eq(batches::bsos.concat(&bsos)))
+        .execute(&db.conn)?;
+    Ok(())
+}
+
+pub fn get(db: &MysqlDb, params: params::GetBatch) -> Result<Option<results::GetBatch>> {
+    let user_id = params.user_id.legacy_id as i32;
+    let collection_id = db.get_collection_id(&params.collection)?;
+    Ok(batches::table
+        .select((batches::bsos, batches::expiry))
+        .filter(batches::user_id.eq(&user_id))
+        .filter(batches::collection_id.eq(&collection_id))
+        .filter(batches::id.eq(&params.id))
+        .get_result(&db.conn)
+        .optional()?)
+}
+
+pub fn delete(db: &MysqlDb, params: params::DeleteBatch) -> Result<()> {
+    let user_id = params.user_id.legacy_id as i32;
+    let collection_id = db.get_collection_id(&params.collection)?;
+    diesel::delete(batches::table)
+        .filter(batches::user_id.eq(&user_id))
+        .filter(batches::collection_id.eq(&collection_id))
+        .filter(batches::id.eq(&params.id))
+        .execute(&db.conn)?;
+    Ok(())
+}
+
+pub fn commit(db: &MysqlDb, params: params::CommitBatch) -> Result<results::CommitBatch> {
+    let bsos = batch_string_to_bsos(&params.batch.bsos)?;
+    db.post_bsos_sync(params::PostBsos {
```

Comment:
Defering the work here to an existing `post_bsos_sync` method should give you a lot of edge-cases for free, such as proper handling of partial updates and ttl bumping.  I like it!  Handling those was a source of some complexity in the python code.